flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -2113,6 +2122,63 @@ IntrinsicLibrary::genRepeat(mlir::Type resultType,
       });
 }
 
+// RESHAPE
+fir::ExtendedValue
+IntrinsicLibrary::genReshape(mlir::Type resultType,
+                             llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 4);
+
+  // Handle source argument
+  auto source = builder.createBox(loc, args[0]);
+
+  // Handle shape argument
+  auto shape = builder.createBox(loc, args[1]);
+  fir::BoxValue shapeTmp = shape;
+  auto shapeRank = shapeTmp.rank();
+  assert(shapeRank == 1);
+  auto shapeTy = shape.getType();
+  auto shapeArrTy = fir::dyn_cast_ptrOrBoxEleTy(shapeTy);
+  auto shapeSeqTy = shapeArrTy.cast<fir::SequenceType>();
+  auto shapeEleTy = shapeSeqTy.getEleTy();
+  assert(fir::isa_integer(shapeEleTy));
+  auto resultRank = shapeSeqTy.getShape();
+
+  // Handle optional pad argument
+  auto pad = isAbsent(args[2])
+                 ? builder.create<fir::AbsentOp>(
+                       loc, fir::BoxType::get(builder.getI1Type()))
+                 : builder.createBox(loc, args[2]);
+
+  // Handle optional order argument
+  auto order = isAbsent(args[3])
+                   ? builder.create<fir::AbsentOp>(
+                         loc, fir::BoxType::get(builder.getI1Type()))
+                   : builder.createBox(loc, args[3]);
+
+  // Create mutable fir.box to be passed to the runtime for the result.
+  auto type = builder.getVarLenSeqTy(resultType, resultRank[0]);
+  auto resultMutableBox =
+      Fortran::lower::createTempMutableBox(builder, loc, type);
+
+  auto resultIrBox =
+      Fortran::lower::getMutableIRBox(builder, loc, resultMutableBox);
+
+  Fortran::lower::genReshape(builder, loc, resultIrBox, source, shape, pad,
+                             order);
+
+  // Handle cleanup of allocatable result descriptor and return
+  auto res = Fortran::lower::genMutableBoxRead(builder, loc, resultMutableBox);
+  return res.match(
+      [&](const fir::ArrayBoxValue &box) -> fir::ExtendedValue {
+        // Add cleanup code
+        addCleanUpForTemp(loc, box.getAddr());
+        return box;
+      },
+      [&](const auto &) -> fir::ExtendedValue {
+        fir::emitFatalError(loc, "unexpected result for RESHAPE");
```

Comment:
Can't reshape also be called on character arrays ?

I still think I badly designed the clean-up in Trim and we should just group the clean-up insertion into:

I think the following should be added as an IntrinsicLibrary member function: 
```
fir::ExtendedValue IntrinsicLibrary::readAndAddCleanUp(fir::MutableBoxValue resultMutableBox) {
   auto res = Fortran::lower::genMutableBoxRead(builder, loc, resultMutableBox);
   return res.match(
     // cover all fir::ArrayBoxValue, fir::CharArrayBoxValue, fir::UnboxedValue, and fir::CharBoxValue cases
   );
}
```

The intrinsic specific fatal error message on unexpected categories brings little in my opinion because we are in control of the result descriptor type and rank. It's seems way more likely to me that we will run into issue by duplicating this pattern and missing to cover some categories.